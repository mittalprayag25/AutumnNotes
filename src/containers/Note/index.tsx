import {Text, Image, View, TouchableOpacity, TextInput} from 'react-native';
import DropDownPicker from 'react-native-dropdown-picker';
import HeaderNav from '../../components/HeaderNav';
import images from '../../assets/index';
import React, {useState} from 'react';
import {useDispatch, useSelector} from 'react-redux';
import actions from '../../redux/actions';
import styles from './styles';

const Note = ({navigation}: any) => {
  const [note, setNote] = useState('');
  const dispatch = useDispatch();
  const {
    categories: {categories},
  }: any = useSelector((s: RootState) => s);

  const [open, setOpen] = useState(false);
  const [value, setValue] = useState('');
  const [items, setItems] = useState(categories ? categories : []);

  const saveNote = () => {
    if (note.length > 0) {
      const noteObj = createNoteObj(note);
      noteObj.category = value;
      dispatch(actions.notes.createnote({note: noteObj}));
      setNote('');
    }
  };

  const createNoteObj = rawNote => {
    const obj: any = {};
    obj.title = rawNote;
    obj.description = rawNote;
    obj.category = '';
    return obj;
  };

  const addNote = item => {
    setNote(item);
  };

  const renderLeftnav = () => (
    <>
      <TouchableOpacity onPress={() => navigation.goBack()} testID="test-back">
        <Image source={images.notes.back} />
      </TouchableOpacity>
    </>
  );

  const renderRightNav = () => (
    <>
      <TouchableOpacity onPress={() => saveNote()} testID="test-save">
        <Image source={images.notes.save} />
      </TouchableOpacity>
    </>
  );

  return (
    <View style={styles.container}>
      <HeaderNav
        leftNav={renderLeftnav()}
        rightNav={renderRightNav()}
        title="Add Notes"
      />

      <View style={styles.container}>
        <View style={styles.editNote}>
          <TextInput
            testID="text-input-notes"
            value={note}
            onChangeText={addNote}
            multiline={true}
            placeholder={'Type here for making notes'}
            style={styles.noteFont}
            selectionColor="#000"
          />
        </View>
        {items.length > 0 && (
          <View>
            <DropDownPicker
              open={open}
              value={value}
              items={items}
              setOpen={setOpen}
              setValue={setValue}
              setItems={setItems}
              placeholder="Choose category"
            />
          </View>
        )}
      </View>
    </View>
  );
};

export default Note;
