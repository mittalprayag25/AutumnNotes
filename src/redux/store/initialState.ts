const initialState = {
  notes: {
    allNotes: [],
    singleNote: null,
  },
  categories: {
    categories: [],
  },
};

export default initialState;
