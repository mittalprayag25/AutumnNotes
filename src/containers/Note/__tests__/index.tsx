import {act, fireEvent, render} from '@testing-library/react-native';
import React from 'react';
import * as reactRedux from 'react-redux';
import {Provider} from 'react-redux';

import {
  createMockStore,
  mockNavigationWithParams,
} from '../../../__mocks__/index';
import Note from '../index';
import initialState from '../../../redux/store/initialState';

const mockedDispatch = jest.fn();

describe('Create Note', () => {
  const navigation = mockNavigationWithParams({});

  const useSelectorSpy = jest.spyOn(reactRedux, 'useSelector');
  const useDispatchSpy = jest.spyOn(reactRedux, 'useDispatch');

  beforeEach(() => {
    useDispatchSpy.mockReturnValue(mockedDispatch);
    useSelectorSpy.mockReturnValue({
      categories: {
        categories: [
          {label: 'cat1', value: 'cat1'},
          {label: 'cat2', value: 'cat2'},
        ],
      },
    });
  });

  const props = {
    navigation: navigation,
  };

  const changeState = {
    ...initialState,
  };
  it('should match snapshot', () => {
    const wrapper = render(
      <Provider store={createMockStore(changeState)}>
        <Note {...props} />
      </Provider>,
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});

describe('Create Node with no categories', () => {
  const navigation = mockNavigationWithParams({});

  const useSelectorSpy = jest.spyOn(reactRedux, 'useSelector');
  const useDispatchSpy = jest.spyOn(reactRedux, 'useDispatch');

  const mockedDispatch = jest.fn();
  useDispatchSpy.mockReturnValue(mockedDispatch);

  beforeEach(() => {
    useDispatchSpy.mockReturnValue(mockedDispatch);
    useSelectorSpy.mockReturnValue({
      categories: {
        categories: [],
      },
    });
  });

  const props = {
    navigation: navigation,
  };

  const changeState = {
    ...initialState,
  };
  it('should match snapshot', () => {
    const wrapper = render(
      <Provider store={createMockStore(changeState)}>
        <Note {...props} />
      </Provider>,
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should save when text input value is there', () => {
    const {getByTestId} = render(
      <Provider store={createMockStore(changeState)}>
        <Note {...props} />
      </Provider>,
    );
    const textInput = getByTestId('text-input-notes');
    const save = getByTestId(`test-save`);
    act(() => {
      fireEvent(save, 'press');
    });
    expect(mockedDispatch).not.toHaveBeenCalled();

    fireEvent.changeText(textInput, '98756789');

    act(() => {
      fireEvent(save, 'press');
    });
    expect(mockedDispatch).toHaveBeenCalled();
  });
  it('should go back', () => {
    const {getByTestId} = render(
      <Provider store={createMockStore(changeState)}>
        <Note {...props} />
      </Provider>,
    );
    const goBack = getByTestId('test-back');
    act(() => {
      fireEvent(goBack, 'press');
    });
    expect(navigation.goBack).toHaveBeenCalledTimes(1);
  });
});
