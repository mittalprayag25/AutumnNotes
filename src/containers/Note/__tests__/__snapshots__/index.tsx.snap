// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Node with no categories should match snapshot 1`] = `
<View
  style={
    Object {
      "height": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderBottomWidth": 1,
        "borderColor": "#abc",
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "space-between",
        "paddingHorizontal": 8,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="test-back"
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/left.png",
            }
          }
        />
      </View>
    </View>
    <Text
      style={
        Object {
          "alignItems": "center",
          "fontSize": 20,
          "fontWeight": "bold",
          "left": "40%",
          "position": "relative",
        }
      }
    >
      Add Notes
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="test-save"
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/save.png",
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "height": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "borderColor": "#eeeee4",
          "borderWidth": 2,
          "height": "80%",
          "marginHorizontal": 16,
          "padding": 16,
        }
      }
    >
      <TextInput
        autoFocus={true}
        multiline={true}
        onChangeText={[Function]}
        placeholder="Type here for making notes"
        selectionColor="#000"
        style={
          Object {
            "fontSize": 20,
          }
        }
        testID="text-input-notes"
        value=""
      />
    </View>
  </View>
</View>
`;

exports[`Create Note should match snapshot 1`] = `
<View
  style={
    Object {
      "height": "100%",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderBottomWidth": 1,
        "borderColor": "#abc",
        "flexDirection": "row",
        "height": 50,
        "justifyContent": "space-between",
        "paddingHorizontal": 8,
        "width": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="test-back"
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/left.png",
            }
          }
        />
      </View>
    </View>
    <Text
      style={
        Object {
          "alignItems": "center",
          "fontSize": 20,
          "fontWeight": "bold",
          "left": "40%",
          "position": "relative",
        }
      }
    >
      Add Notes
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID="test-save"
      >
        <Image
          source={
            Object {
              "testUri": "../../../src/assets/save.png",
            }
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "height": "100%",
      }
    }
  >
    <View
      style={
        Object {
          "borderColor": "#eeeee4",
          "borderWidth": 2,
          "height": "80%",
          "marginHorizontal": 16,
          "padding": 16,
        }
      }
    >
      <TextInput
        autoFocus={true}
        multiline={true}
        onChangeText={[Function]}
        placeholder="Type here for making notes"
        selectionColor="#000"
        style={
          Object {
            "fontSize": 20,
          }
        }
        testID="text-input-notes"
        value=""
      />
    </View>
    <View>
      <View
        style={
          Array [
            Object {
              "width": "100%",
            },
            Object {
              "zIndex": 5000,
            },
            Object {},
          ]
        }
      >
        <View
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          forwardedRef={[Function]}
          onClick={[Function]}
          onLayout={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#fff",
              "borderColor": "#000",
              "borderRadius": 8,
              "borderWidth": 1,
              "flexDirection": "row",
              "justifyContent": "space-between",
              "minHeight": 50,
              "opacity": 1,
              "paddingHorizontal": 10,
              "paddingVertical": 3,
              "width": "100%",
              "zIndex": 5000,
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#000",
                  "flex": 1,
                },
                Object {},
                false,
                Object {},
              ]
            }
          >
            Choose category
          </Text>
          <View
            style={
              Array [
                Object {
                  "marginLeft": 10,
                },
                Object {},
              ]
            }
          >
            <Image
              source={
                Object {
                  "testUri": "../../react-native-dropdown-picker/src/themes/light/icons/arrow-down.png",
                }
              }
              style={
                Array [
                  Object {
                    "height": 20,
                    "width": 20,
                  },
                  Object {},
                ]
              }
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
